### PR TITLE
Addressing remaining issues around handling REST API key

### PR DESCRIPTION
**Pull request by bertold**
These changes update the code so that the REST API key is passed around security inside the integration plugin. The test code had to be updated to use reflection to invoke a constructor in the Secret Java class from Jenkins.

### CODE REVIEW

File: src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java [hunk 1]
```diff
@@ -341,4 +344,16 @@ private void registerPostCreateRetryMappings(
                         .withBody(MablTestConstants.CREATE_DEPLOYMENT_EVENT_RESULT_JSON))
         );
     }
+
+    private Secret mockSecret(String value) {
```
Comment (DonMcNamara) [inline src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java]: Would `Secret.fromString()` work?

File: src/main/java/com/mabl/integration/jenkins/MablStepBuilder.java [hunk 2]
```diff
@@ -245,25 +250,26 @@ public String getDisplayName() {
...
                 @QueryParameter("applicationId") final String applicationId
         ) {
             return MablStepBuilderValidator.validateForm(restApiKey, environmentId, applicationId);
         }
 
-        public ListBoxModel doFillApplicationIdItems(@QueryParameter String restApiKey, @QueryParameter boolean disableSslVerification) {
-            if(restApiKey == null || restApiKey.isEmpty()) {
+        public ListBoxModel doFillApplicationIdItems(@QueryParameter Secret restApiKey, @QueryParameter boolean disableSslVerification) {
+
+            if(restApiKey == null || restApiKey.getPlainText() == null || restApiKey.getPlainText().isEmpty()) {
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablStepBuilder.java]: Seems like logic is being repeated 3x. Could move to a static helper. e.g. `boolean isSecretEmpty(Secret value)` for easier testing.

File: src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java (hunk 1 above)
Comment (twistedpair) [reply src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java]: Not sure you need reflective access here, and I don't think we're returning a mock object, but rather an actual Secret object. Per @DonMcNamara 's comment, can we just mock the method of interest?

Alternatively, `Mockito` may be able to just make a mock for us on it's own, e.g. `mock(Secret.class)`.

File: src/main/java/com/mabl/integration/jenkins/MablStepBuilder.java [hunk 3]
```diff
@@ -52,6 +53,10 @@
  */
 @SuppressWarnings("unused") // automatically discovered by Jenkins
 public class MablStepBuilder extends Builder implements SimpleBuildStep {
+    @Override
+    public String toString() {
```
[the remark this answers is not in the document]
Comment (bertold) [reply src/main/java/com/mabl/integration/jenkins/MablStepBuilder.java]: Good catch! I was debugging something. D'oh.

File: src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java (hunk 1 above)
Comment (bertold) [reply src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java]: Let me try this!

**Comment by bertold** [comment]
> Thanks for working on getting this over the line.
> 
> Curious, do any of these tests TDD the reported finding, which was that the builder form is populated with the secret (which it shouldn't actually receive)?

I need to think about ways to test this assumption. By looking at existing examples, it was not very obvious to me.

File: src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java (hunk 1 above)
Comment (bertold) [reply src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java]: @DonMcNamara Secret.fromString() does not work as it requires Jenkins to be initialized and running, which is not available during test run.

Comment (bertold) [reply src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java]: @twistedpair - unfortunately, mocking does not seem to work as the hudson.util.Secret class is *final*. Mockito will fail. I don't like resorting to reflection, but it works.

**Comment by twistedpair** [comment]
@bertold, Mockito can mock final methods. Did you try setting up MockMaker
<https://github.com/mockito/mockito/wiki/What%27s-new-in-Mockito-2#mock-the-unmockable-opt-in-mocking-of-final-classesmethods>
?

**Comment by bertold** [comment]
I will look at MockMaker for next time around. Mockito alone does not do this.

**Pull request merged**